stable/main.pony

```diff
@@ -8,102 +9,72 @@ actor Main
-
-  fun _print_usage() =>
-    env.out.printv(
-      recover
-        [ "Usage: stable COMMAND [...]"
-          ""
-          "    A simple dependency manager for the Pony language."
-          ""
-          "    Invoke in a working directory containing a bundle.json."
-          ""
-          "Commands:"
-          "    help    - Print this message"
-          "    version - Print version information"
-          "    fetch   - Fetch/update the deps for this bundle"
-          "    env     - Execute the following shell command inside an environment"
-          "              with PONYPATH set to include deps directories. For example,"
-          "              `stable env ponyc myproject`"
-          "    add     - Add a new dependency. For example,"
-          "              `stable add github jemc/pony-inspect"
-          ""
-        ]
-      end)
-
-  fun _load_bundle(create_on_missing: Bool = false): Bundle ? =>
-    let cwd = Path.cwd()
-    match _BundleLocator(env, cwd)
-    | let path: String =>
-        Bundle(FilePath(env.root as AmbientAuth, path)?, log, false)?
-    | None =>
-        if create_on_missing then
-          Bundle(FilePath(env.root as AmbientAuth, cwd)?, log, true)?
-        else
-          log("No bundle.json in current working directory or ancestors.")
-          error
-        end
-    end
+    let cs =
+      try
+        CommandSpec.parent(
+          "stable",
+          """
+          A simple dependency manager for the Pony language.
 
-  fun command_fetch() =>
-    try _load_bundle()?.fetch() end
+          Invoke in a working directory containing a bundle.json.
+          """,
+          [], // options
+          [   // subcommands
+            VersionCmd.command_spec()?
+            FetchCmd.command_spec()?
+            EnvCmd.command_spec()?
+            AddCmd.command_spec()?
+          ])? .> add_help("help", "Shows this text and exits.")?
+      else
+        log("Error instantiating stable cli.")
+        env.exitcode(1)
+        return
+      end
+    let cmd =
+      match CommandParser(cs).parse(env.args, env.vars)
+      | let c: Command => c
+      | let ch: CommandHelp =>
+        ch.print_help(env.out)
+        env.exitcode(0)
+        return
+      | let se: SyntaxError =>
+        log(se.string())
+        log("")
+        Help.general(cs).print_help(env.err)
+        env.exitcode(1)
+        return
+      end
 
-  fun command_env(rest: Array[String] box) =>
-    let ponypath =
+    let locator =
       try
-        let bundle = _load_bundle()?
-        let ponypath' = recover trn String end
-        let iter = bundle.paths().values()
-        let sep = Path.list_sep()(0)?
-        for path in iter do
-          ponypath'.append(path)
-          if iter.has_next() then ponypath'.push(sep) end
-        end
-
-        ponypath'
+        _BundleLocator(env.root as AmbientAuth, log)
       else
-        ""
+        log("unable to locate bundle.json, no sufficient Auth provided.")
```

Comment:
❓ Is this still correct, and as specific as it could be? From what I can tell, [line 50](https://github.com/ponylang/pony-stable/pull/95/files#diff-db754a52c66ac04c799aade5a3e6fa37R50) calls [`create`](https://github.com/ponylang/pony-stable/pull/95/files#diff-555d1573c703b2acd3ac3f86b16581aaR6) and that one can't fail -- or am I missing something?

Reply:
actually, after looking into it, this error should only happen if the `env.root` is `None`, which is not the case if stable is started from cli. This could only happen in testing scenarios when one created a mock `Env`. So, actually this shouldn't happen in day-to-day execution.

Reply:
Ah, ok. Yeah, if you read it "unable" as _"will not be able to locate bundle.json"_, that's actually correct. Let's keep it 😄